python/lsst/daf/butler/script/ingest_files.py

```diff
@@ -0,0 +1,140 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("ingest_files",)
+
+import logging
+from typing import Optional, Tuple, Dict, Any
+from collections import defaultdict
+
+from astropy.table import Table
+
+from lsst.utils import doImport
+
+from .. import Butler, DatasetIdGenEnum
+from ..core import FileDataset, DatasetRef, ButlerURI, DataCoordinate
+
+log = logging.getLogger(__name__)
+
+
+def ingest_files(repo: str, dataset_type: str, run: str, table_file: str,
+                 data_id: Tuple[str, ...] = (),
+                 formatter: Optional[str] = None,
+                 id_generation_mode: str = "UNIQUE",
+                 prefix: Optional[str] = None,
+                 transfer: str = "auto") -> None:
+    """Ingest files from a table.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI string of the Butler repo to use.
+    dataset_type : `str`
+        The name of the dataset type for the files to be ingested.
+    formatter : `str`
+        Fully-qualified python class name for the `Formatter` to use
+        to read the ingested files.
+    run : `str`
+        The run in which the files should be ingested.
+    table_file : `str`
+        Path to a table file to read.
+    id_generation_mode : `str`
+        Mode to use for generating IDs.  Should map to `DatasetGenIdEnum`.
+    prefix : `str`
+        Prefix to use when resolving relative paths in table files. The default
+        is to use the current working directory.
+    transfer : `str`
+        Transfer mode to use for ingest.
+    """
+
+    # Check that the formatter can be imported -- validate this as soon
+    # as possible before we read a potentially large table file.
+    if formatter:
+        doImport(formatter)
+    else:
+        formatter = None
+
+    # Force null prefix to None for API compatibility.
+    if not prefix:
+        prefix = None
+
+    # Convert the dataset ID gen more string to enum
+    id_gen_mode = DatasetIdGenEnum.__members__[id_generation_mode]
+
+    # Create the butler with the relevant run attached.
+    butler = Butler(repo, run=run)
+
+    datasetType = butler.registry.getDatasetType(dataset_type)
+
+    universe = butler.registry.dimensions
+
+    # Convert any additional k=v strings in the dataId tuple to dict
+    # form.
+    common_data_id: Dict[str, Any] = {}
+    for id_str in data_id:
+        dimension_str, value = id_str.split("=")
```

Comment:
You are right that this could be done in the click layer but we got into the habit of passing kwargs straight through from click to the `script` implementation -- that's simpler when iterating on the interface. We also aren't treating the `script` hierarchy as part of the public API so in daf_butler we put the definitions of the subcommands in a single file and then immediately forward to an implementation `script` file. No-one is expected to be calling these apart from click. We always assumed that if we had some reusable functionality in a `script` it would be moved into the shared hierarchy.